pkg/certificateprovider/azurekeyvault/auth.go

```diff
@@ -0,0 +1,164 @@
+		return json.Number(strconv.FormatInt(int64(dur.Round(time.Second).Seconds()), 10))
+	}
+	if _, err := strconv.ParseInt(s, 10, 64); err == nil {
+		// this is the number of seconds case, no conversion required
+		return json.Number(s), nil
+	} else if eo, err := time.Parse(expiresOnDateFormatPM, s); err == nil {
+		return timeToDuration(eo), nil
+	} else if eo, err := time.Parse(expiresOnDateFormat, s); err == nil {
+		return timeToDuration(eo), nil
+	} else {
+		// unknown format
+		return json.Number(""), err
+	}
+}
+
+// Source: https://github.com/Azure/azure-workload-identity/blob/d126293e3c7c669378b225ad1b1f29cf6af4e56d/examples/msal-go/token_credential.go#L25
+func getAADAccessToken(ctx context.Context, tenantID string, clientID string, scope string) (confidential.AuthResult, error) {
+	// Azure AD Workload Identity webhook will inject the following env vars:
+	// 	AZURE_CLIENT_ID with the clientID set in the service account annotation
+	// 	AZURE_TENANT_ID with the tenantID set in the service account annotation. If not defined, then
+	// 	the tenantID provided via azure-wi-webhook-config for the webhook will be used.
+	// 	AZURE_FEDERATED_TOKEN_FILE is the service account token path
+	// 	AZURE_AUTHORITY_HOST is the AAD authority hostname
+
+	tokenFilePath := os.Getenv("AZURE_FEDERATED_TOKEN_FILE")
+	authority := os.Getenv("AZURE_AUTHORITY_HOST")
+
+	if clientID == "" || tokenFilePath == "" || authority == "" {
+		return confidential.AuthResult{}, fmt.Errorf("required environment variables not set, AZURE_CLIENT_ID: %s, AZURE_FEDERATED_TOKEN_FILE: %s, AZURE_AUTHORITY_HOST: %s", clientID, tokenFilePath, authority)
+	}
+
+	// read the service account token from the filesystem
+	signedAssertion, err := readJWTFromFS(tokenFilePath)
+	if err != nil {
+		return confidential.AuthResult{}, errors.Wrap(err, "failed to read service account token")
+	}
+
+	cred, err := confidential.NewCredFromAssertion(signedAssertion)
+	if err != nil {
+		return confidential.AuthResult{}, errors.Wrap(err, "failed to create confidential creds")
+	}
+
+	// create the confidential client to request an AAD token
+	confidentialClientApp, err := confidential.New(
+		clientID,
+		cred,
+		confidential.WithAuthority(fmt.Sprintf("%s%s/oauth2/token", authority, tenantID)))
+	if err != nil {
+		return confidential.AuthResult{}, errors.Wrap(err, "failed to create confidential client app")
+	}
+
+	result, err := confidentialClientApp.AcquireTokenByCredential(ctx, []string{scope})
+	if err != nil {
+		return confidential.AuthResult{}, errors.Wrap(err, "failed to acquire AAD token")
+	}
+
+	return result, nil
+}
+
+// readJWTFromFS reads the jwt from file system
+// Source: https://github.com/Azure/azure-workload-identity/blob/d126293e3c7c669378b225ad1b1f29cf6af4e56d/examples/msal-go/token_credential.go#L88
+func readJWTFromFS(tokenFilePath string) (string, error) {
+	token, err := os.ReadFile(tokenFilePath)
+	if err != nil {
+		return "", err
+	}
+	return string(token), nil
+}
```

Comment:
This is duplicate code that exists in another package

https://github.com/deislabs/ratify/blob/a08ce4e25a10e2b2958957c3ba693d074089e415/pkg/referrerstore/oras/authprovider/azure/azureworkloadidentity.go#L137

https://github.com/deislabs/ratify/blob/a08ce4e25a10e2b2958957c3ba693d074089e415/pkg/referrerstore/oras/authprovider/azure/azureworkloadidentity.go#L180

This can be refactored into a common auth package and reused.

Reply:
factored into common/utils thanks!